Add FakeTensorMode

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #78524
* #78523
* #78522
* #78517
* #78270
* #78090
* #78516
* __->__ #77972
* #77971
* #77970
* #77969

This adds a mode which will intercept calls to `__torch__dispatch__` even if the inputs are not already `FakeTensors`. This mimics the convenient [prior existing usage](https://pytorch.org/torchdistx/latest/fake_tensor.html). It does so by wrapping input tensors to Fake Tensors and then continuing to run the operators. 

Not Yet Implemented:

I still need to memoize conversion of non-fake tensors to fake tensors (and internally, to `meta` devices) following along with the [class here](https://github.com/pytorch/pytorch/blob/master/test/test_meta.py#L70). 

One open question is what should be the duration of the `FakeTensorConverter`. IMO, it would make sense & be convenient for it to live for the duration of `FakeTensorMode`. Since we shouldn't be allocating any new Tensors with actual data (just on `meta` devices) it is probably fine for those tensors to live for the duration of `FakeTensorMode`.

If that is not sufficient, we could try using `weakref.WeakKeyDictionary` mapping tensors to their fake equivalents. I looked into this a bit and there are at least some a few incompatibilities that need to be dealt with.

Commit:
Add FakeTensorMode

[ghstack-poisoned]

Commit:
Update on "Add FakeTensorMode"




[ghstack-poisoned]

Commit:
Update on "Add FakeTensorMode"




[ghstack-poisoned]

Commit:
Update on "Add FakeTensorMode"




[ghstack-poisoned]

Commit:
Update on "Add FakeTensorMode"




[ghstack-poisoned]

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -156,3 +197,6 @@ def merge_devices(t):
         return common_device
+
+class FakeTensorMode(FakeTensor):
+    context = no_dispatch
```
Can we do a modern style mode instead pretty please :)

ezyang replied:
You didn't actually use context, AFAICT?

eellison replied:
What is a modern style mode ? I believe the `context` stopped recursion

ezyang replied:
inherit from TorchDispatchMode

eellison replied:
will do.. mind linking me differences between two why one should do modern style over existing ?

ezyang replied:
the main difference is you can store instance variables on the mode, since it is an actual object not a class

eellison replied:
nice... i probably don't need `setup_mode` then

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -19,6 +21,29 @@
+    # TODO: no real reason to restrict multiple outputs
+    return (
+        len(schema.returns) == 1 and schema.returns[0].type is torch._C.TensorType.get()
+    )
```
Tag opportunity :) cc @anjali411

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -70,22 +108,25 @@ def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+                func, args=args, kwargs=kwargs, normalize_to_only_use_kwargs=True
+            )
+            # cpu is default device if none is specified
+            out_device = new_kwargs.pop("device", torch.device("cpu"))
```
technically it's `torch.get_default_tensor_type()` lol but ok

eellison replied:
which is immutable 😛

Commit:
Update on "Add FakeTensorMode"


This adds a mode which will intercept calls to `__torch__dispatch__` even if the inputs are not already `FakeTensors`. This mimics the convenient [prior existing usage](https://pytorch.org/torchdistx/latest/fake_tensor.html). It does so by wrapping input tensors to Fake Tensors and then continuing to run the operators. 

Not Yet Implemented:

I still need to memoize conversion of non-fake tensors to fake tensors (and internally, to `meta` devices) following along with the [class here](https://github.com/pytorch/pytorch/blob/master/test/test_meta.py#L70). 

One open question is what should be the duration of the `FakeTensorConverter`. IMO, it would make sense & be convenient for it to live for the duration of `FakeTensorMode`. Since we shouldn't be allocating any new Tensors with actual data (just on `meta` devices) it is probably fine for those tensors to live for the duration of `FakeTensorMode`.

If that is not sufficient, we could try using `weakref.WeakKeyDictionary` mapping tensors to their fake equivalents. I looked into this a bit and there are at least some a few incompatibilities that need to be dealt with. 






[ghstack-poisoned]

Commit:
Update on "Add FakeTensorMode"


This adds a mode which will intercept calls to `__torch__dispatch__` even if the inputs are not already `FakeTensors`. This mimics the convenient [prior existing usage](https://pytorch.org/torchdistx/latest/fake_tensor.html). It does so by wrapping input tensors to Fake Tensors and then continuing to run the operators. 

Not Yet Implemented:

I still need to memoize conversion of non-fake tensors to fake tensors (and internally, to `meta` devices) following along with the [class here](https://github.com/pytorch/pytorch/blob/master/test/test_meta.py#L70). 

One open question is what should be the duration of the `FakeTensorConverter`. IMO, it would make sense & be convenient for it to live for the duration of `FakeTensorMode`. Since we shouldn't be allocating any new Tensors with actual data (just on `meta` devices) it is probably fine for those tensors to live for the duration of `FakeTensorMode`.

If that is not sufficient, we could try using `weakref.WeakKeyDictionary` mapping tensors to their fake equivalents. I looked into this a bit and there are at least some a few incompatibilities that need to be dealt with. 






[ghstack-poisoned]

Commit:
Update on "Add FakeTensorMode"


This adds a mode which will intercept calls to `__torch__dispatch__` even if the inputs are not already `FakeTensors`. This mimics the convenient [prior existing usage](https://pytorch.org/torchdistx/latest/fake_tensor.html). It does so by wrapping input tensors to Fake Tensors and then continuing to run the operators. 

Not Yet Implemented:

I still need to memoize conversion of non-fake tensors to fake tensors (and internally, to `meta` devices) following along with the [class here](https://github.com/pytorch/pytorch/blob/master/test/test_meta.py#L70). 

One open question is what should be the duration of the `FakeTensorConverter`. IMO, it would make sense & be convenient for it to live for the duration of `FakeTensorMode`. Since we shouldn't be allocating any new Tensors with actual data (just on `meta` devices) it is probably fine for those tensors to live for the duration of `FakeTensorMode`.

If that is not sufficient, we could try using `weakref.WeakKeyDictionary` mapping tensors to their fake equivalents. I looked into this a bit and there are at least some a few incompatibilities that need to be dealt with. 






[ghstack-poisoned]

Commit:
Update on "Add FakeTensorMode"


This adds a mode which will intercept calls to `__torch__dispatch__` even if the inputs are not already `FakeTensors`. This mimics the convenient [prior existing usage](https://pytorch.org/torchdistx/latest/fake_tensor.html). It does so by wrapping input tensors to Fake Tensors and then continuing to run the operators. 

Not Yet Implemented:

I still need to memoize conversion of non-fake tensors to fake tensors (and internally, to `meta` devices) following along with the [class here](https://github.com/pytorch/pytorch/blob/master/test/test_meta.py#L70). 

One open question is what should be the duration of the `FakeTensorConverter`. IMO, it would make sense & be convenient for it to live for the duration of `FakeTensorMode`. Since we shouldn't be allocating any new Tensors with actual data (just on `meta` devices) it is probably fine for those tensors to live for the duration of `FakeTensorMode`.

If that is not sufficient, we could try using `weakref.WeakKeyDictionary` mapping tensors to their fake equivalents. I looked into this a bit and there are at least some a few incompatibilities that need to be dealt with. 






[ghstack-poisoned]

Commit:
Update on "Add FakeTensorMode"


This adds a mode which will intercept calls to `__torch__dispatch__` even if the inputs are not already `FakeTensors`. This mimics the convenient [prior existing usage](https://pytorch.org/torchdistx/latest/fake_tensor.html). It does so by wrapping input tensors to Fake Tensors and then continuing to run the operators. 

Not Yet Implemented:

I still need to memoize conversion of non-fake tensors to fake tensors (and internally, to `meta` devices) following along with the [class here](https://github.com/pytorch/pytorch/blob/master/test/test_meta.py#L70). 

One open question is what should be the duration of the `FakeTensorConverter`. IMO, it would make sense & be convenient for it to live for the duration of `FakeTensorMode`. Since we shouldn't be allocating any new Tensors with actual data (just on `meta` devices) it is probably fine for those tensors to live for the duration of `FakeTensorMode`.

If that is not sufficient, we could try using `weakref.WeakKeyDictionary` mapping tensors to their fake equivalents. I looked into this a bit and there are at least some a few incompatibilities that need to be dealt with. 






[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36618464).

Commit:
Update on "Add FakeTensorMode"


This adds a mode which will intercept calls to `__torch__dispatch__` even if the inputs are not already `FakeTensors`. This mimics the convenient [prior existing usage](https://pytorch.org/torchdistx/latest/fake_tensor.html). It does so by wrapping input tensors to Fake Tensors and then continuing to run the operators. 

Not Yet Implemented:

I still need to memoize conversion of non-fake tensors to fake tensors (and internally, to `meta` devices) following along with the [class here](https://github.com/pytorch/pytorch/blob/master/test/test_meta.py#L70). 

One open question is what should be the duration of the `FakeTensorConverter`. IMO, it would make sense & be convenient for it to live for the duration of `FakeTensorMode`. Since we shouldn't be allocating any new Tensors with actual data (just on `meta` devices) it is probably fine for those tensors to live for the duration of `FakeTensorMode`.

If that is not sufficient, we could try using `weakref.WeakKeyDictionary` mapping tensors to their fake equivalents. I looked into this a bit and there are at least some a few incompatibilities that need to be dealt with.

Differential Revision: [D36618464](https://our.internmc.facebook.com/intern/diff/D36618464)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36618464).

eellison commented:
I actually think the current behavior makes sense. The easiest mental model IMO is that everything is simulated, and that no real tensors will be affected. I think allowing inplacing views and other things to actually affect the input tensors would be a mistake. 
If you do you would get into a situation where 
`input.t_()` does affect your input, but `input.add_(fake_tensor)` doesn't...

Additionally, the behavior above with `resize_` should just work because the first use of that tensor will get converted to fake, and then any subsequent uses the cached version with the `resize_` applied will be used, which correctly simulates what compute you would have done.

Commit:
Update on "Add FakeTensorMode"


This adds a mode which will intercept calls to `__torch__dispatch__` even if the inputs are not already `FakeTensors`. This mimics the convenient [prior existing usage](https://pytorch.org/torchdistx/latest/fake_tensor.html). It does so by wrapping input tensors to Fake Tensors and then continuing to run the operators. 

Not Yet Implemented:

I still need to memoize conversion of non-fake tensors to fake tensors (and internally, to `meta` devices) following along with the [class here](https://github.com/pytorch/pytorch/blob/master/test/test_meta.py#L70). 

One open question is what should be the duration of the `FakeTensorConverter`. IMO, it would make sense & be convenient for it to live for the duration of `FakeTensorMode`. Since we shouldn't be allocating any new Tensors with actual data (just on `meta` devices) it is probably fine for those tensors to live for the duration of `FakeTensorMode`.

If that is not sufficient, we could try using `weakref.WeakKeyDictionary` mapping tensors to their fake equivalents. I looked into this a bit and there are at least some a few incompatibilities that need to be dealt with.

Differential Revision: [D36618464](https://our.internmc.facebook.com/intern/diff/D36618464)

[ghstack-poisoned]

Commit:
Update on "Add FakeTensorMode"


This adds a mode which will intercept calls to `__torch__dispatch__` even if the inputs are not already `FakeTensors`. This mimics the convenient [prior existing usage](https://pytorch.org/torchdistx/latest/fake_tensor.html). It does so by wrapping input tensors to Fake Tensors and then continuing to run the operators. 

Not Yet Implemented:

I still need to memoize conversion of non-fake tensors to fake tensors (and internally, to `meta` devices) following along with the [class here](https://github.com/pytorch/pytorch/blob/master/test/test_meta.py#L70). 

One open question is what should be the duration of the `FakeTensorConverter`. IMO, it would make sense & be convenient for it to live for the duration of `FakeTensorMode`. Since we shouldn't be allocating any new Tensors with actual data (just on `meta` devices) it is probably fine for those tensors to live for the duration of `FakeTensorMode`.

If that is not sufficient, we could try using `weakref.WeakKeyDictionary` mapping tensors to their fake equivalents. I looked into this a bit and there are at least some a few incompatibilities that need to be dealt with.

Differential Revision: [D36618464](https://our.internmc.facebook.com/intern/diff/D36618464)

[ghstack-poisoned]

Commit:
Update on "Add FakeTensorMode"


This adds a mode which will intercept calls to `__torch__dispatch__` even if the inputs are not already `FakeTensors`. This mimics the convenient [prior existing usage](https://pytorch.org/torchdistx/latest/fake_tensor.html). It does so by wrapping input tensors to Fake Tensors and then continuing to run the operators. 

Not Yet Implemented:

I still need to memoize conversion of non-fake tensors to fake tensors (and internally, to `meta` devices) following along with the [class here](https://github.com/pytorch/pytorch/blob/master/test/test_meta.py#L70). 

One open question is what should be the duration of the `FakeTensorConverter`. IMO, it would make sense & be convenient for it to live for the duration of `FakeTensorMode`. Since we shouldn't be allocating any new Tensors with actual data (just on `meta` devices) it is probably fine for those tensors to live for the duration of `FakeTensorMode`.

If that is not sufficient, we could try using `weakref.WeakKeyDictionary` mapping tensors to their fake equivalents. I looked into this a bit and there are at least some a few incompatibilities that need to be dealt with.

Differential Revision: [D36618464](https://our.internmc.facebook.com/intern/diff/D36618464)

[ghstack-poisoned]

Commit:
Update on "Add FakeTensorMode"


This adds a mode which will intercept calls to `__torch__dispatch__` even if the inputs are not already `FakeTensors`. This mimics the convenient [prior existing usage](https://pytorch.org/torchdistx/latest/fake_tensor.html). It does so by wrapping input tensors to Fake Tensors and then continuing to run the operators. 

Not Yet Implemented:

I still need to memoize conversion of non-fake tensors to fake tensors (and internally, to `meta` devices) following along with the [class here](https://github.com/pytorch/pytorch/blob/master/test/test_meta.py#L70). 

One open question is what should be the duration of the `FakeTensorConverter`. IMO, it would make sense & be convenient for it to live for the duration of `FakeTensorMode`. Since we shouldn't be allocating any new Tensors with actual data (just on `meta` devices) it is probably fine for those tensors to live for the duration of `FakeTensorMode`.

If that is not sufficient, we could try using `weakref.WeakKeyDictionary` mapping tensors to their fake equivalents. I looked into this a bit and there are at least some a few incompatibilities that need to be dealt with.

Differential Revision: [D36618464](https://our.internmc.facebook.com/intern/diff/D36618464)

[ghstack-poisoned]

Commit:
Update on "Add FakeTensorMode"


This adds a mode which will intercept calls to `__torch__dispatch__` even if the inputs are not already `FakeTensors`. This mimics the convenient [prior existing usage](https://pytorch.org/torchdistx/latest/fake_tensor.html). It does so by wrapping input tensors to Fake Tensors and then continuing to run the operators. 

Not Yet Implemented:

I still need to memoize conversion of non-fake tensors to fake tensors (and internally, to `meta` devices) following along with the [class here](https://github.com/pytorch/pytorch/blob/master/test/test_meta.py#L70). 

One open question is what should be the duration of the `FakeTensorConverter`. IMO, it would make sense & be convenient for it to live for the duration of `FakeTensorMode`. Since we shouldn't be allocating any new Tensors with actual data (just on `meta` devices) it is probably fine for those tensors to live for the duration of `FakeTensorMode`.

If that is not sufficient, we could try using `weakref.WeakKeyDictionary` mapping tensors to their fake equivalents. I looked into this a bit and there are at least some a few incompatibilities that need to be dealt with.



[ghstack-poisoned]

eellison commented:
@pytorchbot merge this please

anjali411 commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -20,6 +22,29 @@
+def _is_tensor_constructor(func: OpOverload):
+    assert isinstance(func, OpOverload)
+    schema = func._schema
+    if any(contains_tensor_types(arg.type) for arg in schema.arguments):
```
what about `_like` ops?

eellison replied:
those got added later
